### PR TITLE
Fix some array context mismatching

### PR DESCRIPTION
**Pull request by alexfikl**
Following inducer/arraycontext#22.

Found two issues
* the interleaving code was returning an array with the `_setup_actx`,
* all the caching in `EvaluationMapper` stored thawed data.

Hopefully the CI will catch a few more.

### CODE REVIEW

File: pytential/symbolic/execution.py [hunk 1]
```diff
@@ -281,10 +281,10 @@ def map_common_subexpression(self, expr):
             return self.rec(expr.child)
 
         try:
-            rec = cache[expr.child]
+            rec = thaw(cache[expr.child], self.array_context)
```
[the remark this answers is not in the document]
Comment (inducer) [reply pytential/symbolic/execution.py:284]: IMO: Add an `if` here. What do you think?

Comment (inducer) [inline pytential/symbolic/execution.py:284]: There's little question that things in the discretization cache should be frozen. But the expr-scoped stuff... I think that's overkill.

**Review by inducer — COMMENTED** [review]
Thanks! Two minor things below.

File: pytential/symbolic/execution.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexfikl) [reply pytential/symbolic/execution.py]: Agreed and updated in adcaf93.

Comment (alexfikl) [reply pytential/symbolic/execution.py]: Great catch! Yes, fixed in adcaf93.

**Comment by inducer** [comment]
Thanks!

**Pull request merged**